### PR TITLE
Stale data speedup

### LINKED ISSUE #156: Stale data function check runs slow
@kanderso-nrel and @cwhanse I'm working on integrating some of the pvanalytics functions into the Fleets QA pipeline, and I noticed that the PVanalytics implementation for stale detection is incredibly slow compared to the original pandas version that I put together: 

```
import pvanalytics
import time
import pandas as pd
import matplotlib.pyplot as plt

def stuck_data(time_series):
    time_series = time_series.round(decimals=3)
    #Count consecutive values in the sequence using Pandas functionality
    consecutive_val_counts = time_series.groupby((time_series != time_series.shift()).cumsum()).transform('size') 
    #If 4 or more consecutive values, flag it 
    stuck_data_mask = ~(consecutive_val_counts < 4).astype(bool)
    return stuck_data_mask  

file = "C:/Users/kperry/Documents/source/repos/pvanalytics/pvanalytics/data/ac_power_inv_2173_stale_data.csv"
data = pd.read_csv(file, index_col=0, parse_dates=True)
data = data.asfreq("15T")
data['value_normalized'].plot()

start = time.time()
stuck_mask = pvanalytics.quality.gaps.stale_values_round(data['value_normalized'],
                                                          window=4)
end = time.time()
print("PVAnalytics time execution")
print(end - start)

data['value_normalized'].plot()
data.loc[stuck_mask, "value_normalized"].plot(ls='', marker='.')
plt.legend(labels=["AC Power", "Detected Stale Data"])
plt.xlabel("Date")
plt.ylabel("Normalized AC Power")
plt.title("PVAnalytics Stuck Data mask")
plt.tight_layout()
plt.show()

start = time.time()
stuck_mask_local = stuck_data(data['value_normalized'])
end = time.time()
print("Local time execution")
print(end - start)

data['value_normalized'].plot()
data.loc[stuck_mask_local, "value_normalized"].plot(ls='', marker='.')
plt.legend(labels=["AC Power", "Detected Stale Data"])
plt.xlabel("Date")
plt.ylabel("Normalized AC Power")
plt.title("Local Stuck Data mask")
plt.tight_layout()
plt.show()
```

Running the above outputs the following:

image

With very similar results (my function just counts all stale values in a block as stale, unlike PVAnalytics, which keeps the first stale value):

image

image 

There is obviously a pretty serious change in the runtime for essentially the same function. This difference became really pronounced when I was running hundreds of time series in sequence. I'd recommend we speed this up using some kind of pandas functionality, as it's vectorized.

### PR DESCRIPTION
**Pull request by kperrynrel**
## Description

*Thank you for your contribution! Please provide a brief description of the problem and the proposed solution or new feature (if not already fully described in a linked issue)*

This PR speeds up the stale_data_round() function, by replacing the rolling + apply routine with a vectorized pandas routine.

## Checklist

The following items must be addressed before the code can be merged. 
Please don't hesitate to ask for help if you are unsure of how to accomplish any of the items. 
*You are free to remove any checklist items that do not apply or add additional items that are 
not on this list*

- [X] Closes #156 
- [x] Clearly documented all new API functions with [PEP257](https://www.python.org/dev/peps/pep-0257/) and [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings
- [x] Adds description and name entries in the appropriate "what's new" file 
      in [`docs/whatsnew`](https://github.com/pvlib/pvanalytics/tree/master/docs/whatsnew) 
      for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` 
      or this Pull Request with `` :pull:`num` ``. Includes contributor name 
      and/or GitHub username (link with `` :ghuser:`user` ``).
- [x] Non-API functions clearly documented with docstrings or comments as necessary
- [x] Added tests to cover all new or modified code
- [x] Pull request is nearly complete and ready for detailed review

### CODE REVIEW

**Comment by kperrynrel** [comment]
I tested the new code against the old code to make sure that the results were the same from the old to the new versions of the code, and also tested the code speed up. I ran the following script:

```
import pvanalytics
import pandas as pd
import pathlib
import time

def _backfill_window(endpoints, window):
    # propagate Trues in `endpoints` back `window` periods.  This
    # makes Trues fill the entire window, rather than just marking the
    # right endpoint of each window.
    #
    # `endpoints` must be the output of Series.rolling with `label='right'`
    flags = endpoints
    while window > 0:
        window = window - 1
        flags = flags | endpoints.shift(-window).fillna(False)
    return flags


def _mark(flags, window, mark):
    if mark not in ['all', 'tail', 'end']:
        raise ValueError("mark must be one of'tail', 'end' or 'all'")
    if mark == 'all':
        return _backfill_window(flags, window)
    if mark == 'tail':
        return _backfill_window(flags, window - 1)
    return flags

def stale_values_round_original(x, window=6, decimals=3, mark='tail'):
    """Identify stale values by rounding.
    A value is considered stale if it is part of a sequence of length
    `window` of values that are identical when rounded to `decimals`
    decimal places.
    Parameters
    ----------
    x : Series
        Data to be processed.
    window : int, default 6
        Number of consecutive identical values for a data point to be
        considered stale.
    decimals : int, default 3
        Number of decimal places to round to.
    mark : str, default 'tail'
        How much of the window to mark ``True`` when a sequence of
        stale values is detected. Can be one of 'tail', 'end', or
        'all'.
        - If 'tail' (the default) then every point in the window
          *except* the first point is marked ``True``.
        - If 'end' then the first `window - 1` values in a stale
          sequence sequence are marked ``False`` and all subsequent
          values in the sequence are marked ``True``.
        - If 'all' then every point in the window *including* the
          first point is marked ``True``.
    Returns
    -------
    Series
        True for each value that is part of a stale sequence of data.
    Raises
    ------
    ValueError
        If `mark` is not one of 'tail', 'end', or 'all'.
    Notes
    -----
    Based on code from the pvfleets_qa_analysis project. Copyright
    (c) 2020 Alliance for Sustainable Energy, LLC.
    """
    rounded_diff = x.round(decimals=decimals).diff()
    endpoints = rounded_diff.rolling(window=window-1).apply(
        lambda xs: len(xs[xs == 0]) == window-1
    ).fillna(False).astype(bool)
    return _mark(endpoints, window, mark)

def stale_values_round(x, window=6, decimals=3, mark='tail'):
    """Identify stale values by rounding.

    A value is considered stale if it is part of a sequence of length
    `window` of values that are identical when rounded to `decimals`
    decimal places.

    Parameters
    ----------
    x : Series
        Data to be processed.
    window : int, default 6
        Number of consecutive identical values for a data point to be
        considered stale.
    decimals : int, default 3
        Number of decimal places to round to.
    mark : str, default 'tail'
        How much of the window to mark ``True`` when a sequence of
        stale values is detected. Can be one of 'tail', 'end', or
        'all'.

        - If 'tail' (the default) then every point in the window
          *except* the first point is marked ``True``.
        - If 'end' then the first `window - 1` values in a stale
          sequence sequence are marked ``False`` and all subsequent
          values in the sequence are marked ``True``.
        - If 'all' then every point in the window *including* the
          first point is marked ``True``.

    Returns
    -------
    Series
        True for each value that is part of a stale sequence of data.

    Raises
    ------
    ValueError
        If `mark` is not one of 'tail', 'end', or 'all'.

    Notes
    -----
    Based on code from the pvfleets_qa_analysis project. Copyright
    (c) 2020 Alliance for Sustainable Energy, LLC.

    """
    rounded_x = x.round(decimals=decimals)
    consecutive_val_counts = rounded_x.groupby((rounded_x !=
                                                rounded_x.shift())
                                               .cumsum()).cumcount()
    endpoints = (~(consecutive_val_counts < window-1)) & ~(rounded_x.isna())
    return _mark(endpoints, window, mark)


pvanalytics_dir = pathlib.Path(pvanalytics.__file__).parent
file = pvanalytics_dir / 'data' / 'ac_power_inv_2173_stale_data.csv'
data = pd.read_csv(file, index_col=0, parse_dates=True)
data = data.asfreq("15T")

# TAIL
# Try the original slower logic
start = time.time()
stale_mask_old = stale_values_round_original(data['value_normalized'])
end = time.time()

print("Old function run time (s)")
print(end-start)

# Try the new logic
start = time.time()
stale_mask_new = stale_values_round(data['value_normalized'])
end = time.time()

print("New function run time (s)")
print(end-start)

# Compare the results
print("Stale mask new-old comparison: mark=tail")
print(all(stale_mask_old==stale_mask_new))

# END
# Try the original slower logic
start = time.time()
stale_mask_old = stale_values_round_original(data['value_normalized'], mark='end')
end = time.time()

print("Old function run time (s)")
print(end-start)

# Try the new logic
start = time.time()
stale_mask_new = stale_values_round(data['value_normalized'], mark='end')
end = time.time()

print("New function run time (s)")
print(end-start)

# Compare the results
print("Stale mask new-old comparison: mark=end")
print(all(stale_mask_old==stale_mask_new))

# ALL
# Try the original slower logic
start = time.time()
stale_mask_old = stale_values_round_original(data['value_normalized'], mark='all')
end = time.time()

print("Old function run time (s)")
print(end-start)

# Try the new logic
start = time.time()
stale_mask_new = stale_values_round(data['value_normalized'], mark='all')
end = time.time()

print("New function run time (s)")
print(end-start)

# Compare the results
print("Stale mask new-old comparison: mark=all")
print(all(stale_mask_old==stale_mask_new))

```

Here are the results, which show consistent speedup+ the same results:

image

File: pvanalytics/quality/gaps.py [hunk 1]
```diff
@@ -119,7 +114,6 @@ def stale_values_diff(x, window=6, rtol=1e-5, atol=1e-8, mark='tail'):
     """
     if window < 2:
         raise ValueError('window set to {}, must be at least 2'.format(window))
-
```
Comment (kanderso-nrel) [inline pvanalytics/quality/gaps.py:122]: Were all these blank lines removed intentionally?  I think I'd keep all of them in, personally

**Review by kanderso-nrel — APPROVED** [review]
OK with me

File: pvanalytics/quality/gaps.py (hunk 1 above)
Comment (kperrynrel) [reply pvanalytics/quality/gaps.py]: That was a mistake on my part! Added those guys back in

**Comment by wholmgren** [comment]
We would use this today if it were merged and released :) The old function is a significant bottleneck in a workflow.

**Comment by kanderso-nrel** [comment]
Having taken a second look, how about this?  Seems about as fast as the proposed code, but easier to grok?

```python
    rounded_diff = x.round(decimals=decimals).diff()
    endpoints = (rounded_diff == 0).rolling(window=window-1).sum() == window-1
    return _mark(endpoints, window, mark)
```

**Comment by kperrynrel** [comment]
Hey all, I subbed in @kanderso-nrel's code and ran a comparison:

image

All the results line up when comparing this new code to the original implemented code and the speed-up is essentially identical so I'm for using this version as well because it's cleaner. Thanks for the tip @kanderso-nrel !

File: docs/whatsnew/0.1.3.rst [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 Enhancements
 ~~~~~~~~~~~~
+* Updated the :py:func:`~pvanalytics.gaps.stale_values_round` function with pandas groupby() functionality, leading to the same results with a 300X speedup. (:issue:`156`, :pull:`158`)
```
Comment (cwhanse) [inline docs/whatsnew/0.1.3.rst:8]:
```suggestion
* Updated the :py:func:`~pvanalytics.gaps.stale_values_round` function with pandas functionality, leading to the same results with a 300X speedup. (:issue:`156`, :pull:`158`)
```

**Review by cwhanse — APPROVED** [review]

**Comment by wholmgren** [comment]
Thanks all!

**Pull request merged**